src/org/zaproxy/zap/extension/jsonView/HttpPanelJsonView.java

```diff
@@ -0,0 +1,174 @@
+		//TODO disable line numbers, enable folding (requires rsyntaxtextview 2.6.1)
+		httpPanelJsonArea = new HttpPanelJsonArea();
+		RTextScrollPane scrollPane = new RTextScrollPane(httpPanelJsonArea);
+		scrollPane.setLineNumbersEnabled(false);
+		mainPanel = new JPanel(new BorderLayout());
+		mainPanel.add(scrollPane, BorderLayout.CENTER);
+		this.model = model;
+		model.addHttpPanelViewModelListener(this);
+	}
+
+	@Override
+	public void setSelected(boolean selected) {
+		if (selected) {
+			httpPanelJsonArea.requestFocusInWindow();
+		}
+	}
+
+	@Override
+	public String getName() {
+		return NAME;
+	}
+
+	@Override
+	public String getCaptionName() {
+		return CAPTION_NAME;
+	}
+
+	@Override
+	public String getTargetViewName() {
+		return "";
+	}
+
+	@Override
+	public int getPosition() {
+		return 0;
+	}
+
+	@Override
+	public boolean isEnabled(Message message) {
+		if(message == null){
+			return false;
+		}
+		if(message instanceof HttpMessage){
+			HttpHeader header;
+			if(this.model instanceof  RequestBodyStringHttpPanelViewModel){
+				header = ((HttpMessage) message).getRequestHeader();
+			} else if (this.model instanceof ResponseBodyStringHttpPanelViewModel){
+				header = ((HttpMessage) message).getResponseHeader();
+			} else {
+				return false;
+			}
+			if(header.getHeader("Content-Type") != null){
+				return header.getHeader("Content-Type").trim().contains("application/json");
+			}
+			//Sometimes the content-type is wrong, then this is a shame
+			return false;
+		}
+		//WebSocket messages? why not
+		return true;
+	}
+
+	@Override
+	public boolean hasChanged() { return true; }
+
+	@Override
+	public JComponent getPane() {
+		return mainPanel;
+	}
+
+	@Override
+	public boolean isEditable() {
+		return httpPanelJsonArea.isEditable();
+	}
+
+	@Override
+	public void setEditable(boolean editable) {
+		httpPanelJsonArea.setEditable(editable);
+	}
+
+	@Override
+	public HttpPanelViewModel getModel() {
+		return model;
+	}
+
+	@Override
+	public void save() {
+		//we intentionally want to let the user give broken data
+		this.model.setData(httpPanelJsonArea.getText());
+	}
+
+	@Override
+	public void setParentConfigurationKey(String configurationKey) {
+	}
+
+	@Override
+	public void loadConfiguration(FileConfiguration fileConfiguration) {
+	}
+
+	@Override
+	public void saveConfiguration(FileConfiguration fileConfiguration) {
+	}
+
+	@Override
+	public void dataChanged(HttpPanelViewModelEvent e) {
+		String body = ((AbstractStringHttpPanelViewModel) e.getSource()).getData();
+		//TODO setup strict parsing
+		JSON object;
+		try{
+			if(body.startsWith("{")){
+				object = JSONObject.fromObject(body);
+			} else if (body.startsWith("[")){
+				object = JSONArray.fromObject(body);
+			} else{
+				throw new JSONException("Expected a '{' or '[' at character 0");
+			}
+			httpPanelJsonArea.setText(object.toString(2));
+		} catch(JSONException ex){
+			httpPanelJsonArea.setText("invalid JSON");

```

Comment:
Title caps?

If this is catching the exception thrown inside the try wouldn't you want the actual exception message? (Otherwise why even bother customizing the message 4 lines earlier?)

Reply:
I got a friend to check this out and actually there are some issues with setting an error message as the body, since save() will overwrite the actual data in the message model and the original will be lost. Needs some bigger reworking too...

Reply:
It could be used a flag for the error state, if there's an error don't save.

Reply:
(And should also disable the text area.)

Reply:
I've been evaluating this, on one side I want to be able to save broken data because sometimes that's what's needed to make something interesting in the backend. On the other side the view should mainly be used for working with json, which is mostly valid in messages.

My idea is to view the data unedited if it is invalid, and indented if it is valid. However, with new messages the json view will only be enabled if parsing of an unedited message succeeds. This way switching between views a user can break the json syntax without any consequences. There are a lot of unintended cases when the view is useful (text/plain, application/x-ipynb+json, etc.), so I had to rewrite the enabler and selector pretty much.

About casing, is Json in title caps _everywhere_ fine? Also in addon description etc.